File: 0101-0200/0149.Max_Points_on_a_Line.py
#27ms






class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        if len(points) <= 2:
            return len(points)
        
        def find_slope(p1, p2):
            x1, y1 = p1
            x2, y2 = p2
            if x1-x2 == 0:
                return inf
            return (y1-y2)/(x1-x2)
        
        ans = 1
        for i, p1 in enumerate(points):
            slopes = defaultdict(int)
            for j, p2 in enumerate(points[i+1:]):
                slope = find_slope(p1, p2)
                slopes[slope] += 1
                ans = max(slopes[slope], ans)
        return ans+1






#19ms







class Solution:
    def maxPoints(self, points: list[list[int]]) -> int:
                                                #   points = [[1,0],[2,1],[3,4], [5,4]]

        points.sort()                           #   point1 point2 (dx,dy)    m     slope                M
        slope, M = defaultdict(int), 0          #   –––––– –––––– ––––––– –––––––  –––––––––––        –––––
                                                #   [1,0]  [2,1]   (1,1)   (1,1)   {(1,1):1}            1
        for i, (x1, y1) in enumerate(points):   #          [3,4]   (2,4)   (1,2)   {(1,1):1,(1,2):1}    1
                                                #          [5,4]   (4,4)   (1,1)   {(1,1):2,(1,2):1}    2
            slope.clear()                       #   [2,1]  [3,4]   (1,3)   (1,3)   {(1,3):1}            2
                                                #          [5,4]   (3,3)   (1,1)   {(1,3):1,(1,1):1}    2
            for x2, y2 in points[i + 1:]:       #   [3,4]  [5,4]   (3,0)   (1,0)   {(1,0):1}            2
                dx, dy = x2 - x1, y2 - y1
                                                #  M + 1 = 2 + 1 = 3 <-- return
                G = gcd(dx, dy)                 
                m = (dx//G,dy//G)
                
                slope[m] += 1
                if slope[m] > M: M = slope[m]
    
        return M + 1

File: 0101-0200/0149.Max_Points_on_a_Line.cpp
//37ms





class Solution {
public:
    int maxPoints(vector<vector<int>>& points) {
        int maxi = 0;
        // x -> x[0],x[1]  (x1,y1)
        for (auto& x : points) {
            map<double, int> slopes;
            // y -> y[0],y[1]  (x2,y2)
            for (auto& y : points) {
                if (x == y) continue;
                double slope = numeric_limits<double>::infinity();
                // if (x2 - x1 != 0)
                if (y[0] - x[0] != 0) {
                    // slope = (y2 - y1) / (x2 - x1)
                    slope = (double)(y[1] - x[1]) / (y[0] - x[0]);
                }
                ++slopes[slope];
                maxi = max(maxi, slopes[slope]);
            }
        }
        return maxi + 1;
    }
};






//6ms






class Solution {
public:
    int gcd(int a, int b) {
        if (b==0)
            return a;
        return gcd(b, a%b);
    }

    int maxPoints(vector<vector<int>>& points) {
        // map<pair<pair<int, int>, int>, int> lines;
        vector<pair<int, int>> slopes;
        int a, b, cnt, _gcd, n=points.size(), max_points=1;

        for(int i=0;i<n;i++) {
            slopes.clear();

            for(int j=i+1;j<n;j++){
                a = points[j][1] - points[i][1];
                b = points[j][0] - points[i][0];
                _gcd = gcd(a, b);
                a /= _gcd;
                b /= _gcd;
                
                slopes.push_back({a, b});
            }
            sort(slopes.begin(), slopes.end());

            cnt = 0;
            for(int x=0;x<slopes.size();x++) {
                ++cnt;
                if(x==slopes.size()-1 || slopes[x]!=slopes[x+1]) {
                    max_points = max(max_points, cnt+1);
                    cnt = 0;
                }
            }
        }


        return max_points;
    }
};

